File: packages/shared/src/research/research-news-sources.ts
/**
 * Research news-source opt-in/out pack + merge.
 *
 * Source of truth for which `rss:{feed_id}` news feeds are active, shared by
 * worker ingest, daily-report build, and the research hub. Feeds are grouped by
 * topic/brand in `config/research_news_sources.yaml`; feed URLs stay in
 * `config/config.yaml` (never duplicated here).
 *
 * Default-ON / opt-out: empty workspace overlay => every default group ON.
 * Precedence: masterEnabled=false > excludedFeeds > enabledFeeds > group > defaults.
 *
 * NOTE: no node:fs / node:path here — the seed YAML loader lives in
 * `./research-news-sources-seed.ts` (Node-only) so this module stays safe for
 * Vite client bundling and Convex edge-runtime sync.
 */

export type NewsSourceGroup = {
  id: string;
  label: string;
  feeds: string[];
};

export type NewsSourcesSeed = {
  version: string;
  groups: NewsSourceGroup[];
  /** Feed ids in catalog order (unique across all groups). */
  catalogIds: string[];
  /** Group ids defaulted ON (opt-out). */
  defaultGroupIds: string[];
};

export type NewsSourcesWorkspaceValue = {
  version: 1;
  /** Absent/true => master ON. Explicit false => effective []. */
  masterEnabled?: boolean;
  /** Group ids the desk turned OFF. */
  excludedGroups: string[];
  /** Feed ids the desk turned OFF (wins over group membership). */
  excludedFeeds: string[];
  /** Feed ids force-ON inside an excluded group (escape hatch). */
  enabledFeeds: string[];
};

function parseGroup(raw: unknown, label: string): NewsSourceGroup {
  if (!raw || typeof raw !== "object") {
    throw new Error(`Invalid news-source group in ${label}`);
  }
  const row = raw as Record<string, unknown>;
  const id = typeof row.id === "string" ? row.id.trim() : "";
  const groupLabel = typeof row.label === "string" ? row.label.trim() : "";
  if (!id || !groupLabel) {
    throw new Error(`News-source group in ${label} requires id and label`);
  }
  const feeds = Array.isArray(row.feeds)
    ? row.feeds.filter((f): f is string => typeof f === "string").map((f) => f.trim())
    : [];
  if (feeds.length === 0) {
    throw new Error(`News-source group ${id} requires at least one feed`);
  }
  return { id, label: groupLabel, feeds };
}

export function parseResearchNewsSourcesSeed(doc: unknown): NewsSourcesSeed {
  if (!doc || typeof doc !== "object") {
    throw new Error("News-sources seed must be an object");
  }
  const root = doc as Record<string, unknown>;
  const version =
    typeof root.version === "string" && root.version.trim() ? root.version.trim() : "v1";
  const groups = (Array.isArray(root.groups) ? root.groups : []).map((g, i) =>
    parseGroup(g, `groups[${i}]`),
  );
  if (groups.length === 0) {
    throw new Error("News-sources seed requires at least one group");
  }

  const catalogIds: string[] = [];
  const seen = new Set<string>();
  for (const group of groups) {
    for (const feed of group.feeds) {
      if (seen.has(feed)) {
        throw new Error(`Duplicate news-source feed id in catalog: ${feed}`);
      }
      seen.add(feed);
      catalogIds.push(feed);
    }
  }

  const defaultsRoot = (root.defaults as Record<string, unknown> | undefined) ?? {};
  const defaultGroupIds = Array.isArray(defaultsRoot.groups)
    ? defaultsRoot.groups.filter((id): id is string => typeof id === "string")
    : [];
  if (defaultGroupIds.length === 0) {
    throw new Error("News-sources seed defaults.groups must be non-empty");
  }
  const groupIdSet = new Set(groups.map((g) => g.id));
  for (const id of defaultGroupIds) {
    if (!groupIdSet.has(id)) {
      throw new Error(`News-sources defaults references unknown group: ${id}`);
    }
  }

  return { version, groups, catalogIds, defaultGroupIds };
}

export function emptyNewsSourcesWorkspace(): NewsSourcesWorkspaceValue {
  return { version: 1, excludedGroups: [], excludedFeeds: [], enabledFeeds: [] };
}

/**
 * Parse a raw workspace_config value into the opt-out shape.
 * Missing / empty row => default (master ON, no excludes).
 */
export function parseNewsSourcesWorkspace(raw: unknown): NewsSourcesWorkspaceValue {
  if (!raw || typeof raw !== "object") {
    return emptyNewsSourcesWorkspace();
  }
  const row = raw as Record<string, unknown>;
  const asList = (v: unknown): string[] =>
    Array.isArray(v) ? v.filter((x): x is string => typeof x === "string") : [];
  const masterEnabled =
    typeof row.masterEnabled === "boolean" ? row.masterEnabled : undefined;
  return {
    version: 1,
    ...(masterEnabled === undefined ? {} : { masterEnabled }),
    excludedGroups: asList(row.excludedGroups),
    excludedFeeds: asList(row.excludedFeeds),
    enabledFeeds: asList(row.enabledFeeds),
  };
}

/**
 * Effective active feed-id set (catalog order), opt-out semantics.
 * master off => []. Never returns empty when master is on and the seed has feeds
 * (a bad exclude-all falls back to the seed defaults).
 */
export function mergeNewsSources(
  seed: NewsSourcesSeed,
  workspace: NewsSourcesWorkspaceValue,
): string[] {
  if (workspace.masterEnabled === false) return [];

  const groupIdSet = new Set(seed.groups.map((g) => g.id));
  const excludedGroups = new Set(workspace.excludedGroups.filter((id) => groupIdSet.has(id)));
  const defaultGroupSet = new Set(seed.defaultGroupIds);

  // base = feeds of default groups that are not explicitly excluded, in catalog order.
  // (opt-out: only `defaultGroupIds` are ON by default; every other group is OFF
  //  unless a feed is force-enabled via `enabledFeeds`.)
  const baseSet = new Set<string>();
  const base: string[] = [];
  for (const group of seed.groups) {
    if (!defaultGroupSet.has(group.id)) continue; // not a default -> off unless force-enabled
    if (excludedGroups.has(group.id)) continue;
    for (const feed of group.feeds) {
      if (baseSet.has(feed)) continue;
      baseSet.add(feed);
      base.push(feed);
    }
  }

  // escape hatch: enabledFeeds force-ON (feed-level), even inside an excluded/default-off group.
  const catalogSet = new Set(seed.catalogIds);
  const enabledSet = new Set(workspace.enabledFeeds.filter((f) => catalogSet.has(f)));
  const excludedSet = new Set(workspace.excludedFeeds);

  const effective: string[] = [];
  const effectiveSet = new Set<string>();
  const push = (feed: string) => {
    if (effectiveSet.has(feed)) return;
    effectiveSet.add(feed);
    effective.push(feed);
  };
  for (const feed of base) {
    if (excludedSet.has(feed)) continue;
    push(feed);
  }
  for (const feed of enabledSet) {
    if (excludedSet.has(feed)) continue; // excludedFeeds wins
    push(feed);
  }

  if (effective.length === 0 && seed.catalogIds.length > 0) {
    return [...seed.catalogIds];
  }
  return effective;
}

File: apps/worker/research_news_sources.py
# coding=utf-8
"""
Research news-source opt-in/out twin (Python).

Mirror of packages/shared/src/research/research-news-sources.ts so the WORKER
(offline, direct-Convex) can compute the same effective `rss:{feed_id}` set as
the daily-report build + hub. Reads config/research_news_sources.yaml, merges the
workspace overlay (`research.enabledNewsSources`) opt-out shape, and filters the
catalog ids. Non-catalog feeds are untouched.
"""

from __future__ import annotations

from pathlib import Path
from typing import Dict, List, Optional

import yaml


def default_news_sources_path() -> Path:
    return Path(__file__).resolve().parents[2] / "config" / "research_news_sources.yaml"


def load_news_sources_seed(path: Optional[Path] = None) -> Dict:
    """Group-id -> [feed-ids] + defaultGroupIds, in YAML order (dict preserves order)."""
    p = path or default_news_sources_path()
    with open(p, encoding="utf-8") as handle:
        doc = yaml.safe_load(handle) or {}
    groups_meta: List[tuple] = []
    for g in doc.get("groups", []):
        groups_meta.append((g.get("id"), g.get("feeds", [])))
    defaults = (doc.get("defaults") or {}).get("groups", []) or []
    return {
        "groups": groups_meta,  # [(group_id, [feed,...])]
        "defaultGroupIds": defaults,
    }


def _as_list(raw, key) -> List[str]:
    val = raw.get(key) if isinstance(raw, dict) else None
    if not isinstance(val, list):
        return []
    return [str(x).strip() for x in val if isinstance(x, str) and x.strip()]


def empty_news_sources_workspace() -> Dict:
    return {"masterEnabled": True, "excludedGroups": [], "excludedFeeds": [], "enabledFeeds": []}


def parse_news_sources_workspace(raw) -> Dict:
    if not isinstance(raw, dict):
        return empty_news_sources_workspace()
    master = raw.get("masterEnabled")
    return {
        "masterEnabled": master if isinstance(master, bool) else True,
        "excludedGroups": _as_list(raw, "excludedGroups"),
        "excludedFeeds": _as_list(raw, "excludedFeeds"),
        "enabledFeeds": _as_list(raw, "enabledFeeds"),
    }


def catalog_feed_ids(seed: Dict) -> List[str]:
    out: List[str] = []
    for _, feeds in seed["groups"]:
        for f in feeds:
            if f not in out:
                out.append(f)
    return out


def merge_news_sources(seed: Dict, workspace: Dict) -> List[str]:
    """Effective active feed-id set (catalog order), opt-out semantics.

    master off => []. Bad exclude-all with master on => full catalog (never empty).
    """
    if workspace.get("masterEnabled") is False:
        return []

    default_group_set = set(seed["defaultGroupIds"])
    excluded_groups = set(workspace.get("excludedGroups", []))
    base: List[str] = []
    base_set: set = set()
    for group_id, feeds in seed["groups"]:
        if group_id not in default_group_set:
            continue
        if group_id in excluded_groups:
            continue
        for f in feeds:
            if f in base_set:
                continue
            base_set.add(f)
            base.append(f)

    catalog_set = set(catalog_feed_ids(seed))
    enabled_set = {f for f in workspace.get("enabledFeeds", []) if f in catalog_set}
    excluded_set = set(workspace.get("excludedFeeds", []))

    effective: List[str] = []
    effective_set: set = set()
    for f in base:
        if f in excluded_set:
            continue
        effective_set.add(f)
        effective.append(f)
    for f in enabled_set:
        if f in excluded_set:
            continue
        if f in effective_set:
            continue
        effective_set.add(f)
        effective.append(f)

    if not effective and catalog_set:
        return catalog_feed_ids(seed)
    return effective
